File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:

    vector<pair<int,int>> getNeighbours(pair<int,int> node,vector<vector<int>>& grid)
    {
        vector<pair<int,int>> ans;
        vector<pair<int,int>> coords = {{-1,0},{0,-1},{0,1},{1,0}};
        for(int i=0;i<4;i++)
        {
            int x = node.first + coords[i].first;
            int y = node.second + coords[i].second;
            if(x>=0 && y>=0 && x<grid.size() && y<grid[0].size() && grid[x][y]==1) ans.push_back({x,y});
        }
        return ans;
    }

    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<int,int>> q;
        for(int i=0;i<grid.size();i++)
        {
            for(int j=0;j<grid[0].size();j++)
            {
                if(grid[i][j]==2) q.push({i,j});
            }
        }

        //multi-source BFS
        int time = 0;
        bool flag=false;
        while(!q.empty())
        {
            int stepSize = q.size();
            while(stepSize--)
            {
                pair<int,int> node = q.front();
                q.pop();
                for(auto n : getNeighbours(node,grid))
                {
                    flag=true;
                    q.push(n);
                    grid[n.first][n.second]=2;
                }
            }
            if(flag)
            {
                time++;
                flag= !flag;
            }
        }

        for(int i=0;i<grid.size();i++)
        {
            for(int j=0;j<grid[0].size();j++)
            {
                if(grid[i][j]==1) return -1;
            }
        }
        return time;
    }
};

File: 1036-rotting-oranges/rotting-oranges.java
class Pair{
    int x;
    int y;

    public Pair(int x,int y){
        this.x = x;
        this.y =y;
    }
}

class Solution {


    public ArrayList<Pair> getNeighbours(Pair pos, int[][] grid,int m,int n){
        ArrayList<Pair> ans = new ArrayList<>();
        int i = pos.x;
        int j = pos.y;
        if(i-1>=0 && i-1<m && j>=0 && j<n && grid[i-1][j]==1) ans.add(new Pair(i-1,j));
        if(i>=0 && i<m && j-1>=0 && j-1<n && grid[i][j-1]==1) ans.add(new Pair(i,j-1));
        if(i>=0 && i<m && j+1>=0 && j+1<n && grid[i][j+1]==1) ans.add(new Pair(i,j+1));
        if(i+1>=0 && i+1<m && j>=0 && j<n && grid[i+1][j]==1) ans.add(new Pair(i+1,j));
        return ans;
    }

    public int orangesRotting(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        Deque<Pair> rotten = new ArrayDeque<>();
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(grid[i][j]==2) rotten.addLast(new Pair(i,j));
            }
        }
        int minutes = 0;
        while(rotten.size()>0)
        {
            int levelsize = rotten.size();
            while(levelsize>0)
            {
                Pair pos = rotten.removeFirst();
                for(var nb : getNeighbours(pos,grid,m,n))
                {
                    grid[nb.x][nb.y] = 2;
                    rotten.addLast(nb);
                }
                levelsize--;
            }
            minutes++;
        }
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(grid[i][j]==1) return -1;
            }
        }
        return (minutes==0) ? 0 : minutes-1;

    }
}